### PR TITLE
contains-own added

### PR DESCRIPTION
**Pull request by sm-ghavami**
contains-own pseudo selector added to the project

### CODE REVIEW

**Comment by facelessuser** [comment]
For the most part, we stick to the CSS spec, though there are maybe 2 cases where we've deviated from this: `:contains()`, and `[attr!=value]`. We are not the only ones to have used such selectors, and they are fairly useful, so I made an exception. What I'd like to avoid though is going to far down the road of having numerous custom pseudo classes.

If we did go down this road, I may start considering some kind of prefix like webrowsers use: `:-soup-contains-own`. This would mainly be to prevent clashing with future pseudo classes. I would need to think about whether I want to go down this road though.

I will at least acknowledge why one would find this feature useful, I'm just not sure yet if this is the path I want to go down yet.

**Comment by facelessuser** [comment]
Overall, the code looks fine. I don't think we needed to duplicate all of the "contains" tests though. I would narrow them down to what is needed and most useful.

**Comment by facelessuser** [comment]
I would need to probably update Python 3.9 testing as the CI seems to be having some trouble with how it is currently configured.

**Comment by facelessuser** [comment]
@sm-ghavami, let me ask, is this a completely unique solution, or one that is found in other CSS libraries?

**Comment by sm-ghavami** [comment]
> @sm-ghavami, let me ask, is this a completely unique solution, or one that is found in other CSS libraries?

I have worked with a lot of libraries in different programming languages to scrape information from xml/html/... and I was comfortable with jsoup (Java language) as it has a complete list of selectors to extract what you want. So I edited the beautifulsoup library to have some necessary selectors such as contains, containsOwn,... (you can find the library in my repositories list).
As I had my library in recent years, I hadn't taken a look at the new versions of the beautifulsoup library. Today I just took a look at the library to see the changes and I found that the css selector is replaced by your library. After reading your library I found it pretty well formed. So I decided to have a contribution to this project. In my experience, contains-own is as essential as contains selector. So I decided to add the selector to the project. 

It is very interesting to have a prefix/postfix to create a general solution for the problem.

**Comment by facelessuser** [comment]
> It is very interesting to have a prefix/postfix to create a general solution for the problem.

Interesting bad or good? The idea is to prevent any possible conflict in the future with new official CSS selectors. This is much how browsers handle non-standard pseudo classes.

**Comment by sm-ghavami** [comment]
> > It is very interesting to have a prefix/postfix to create a general solution for the problem.
> 
> Interesting bad or good? The idea is to prevent any possible conflict in the future with new official CSS selectors. This is much how browsers handle non-standard pseudo classes.

The idea is good if we let the prefix/postfix to be applied to small group of selectors and expand it if there is no problem with the expansion.

**Comment by facelessuser** [comment]
> and expand it if there is no problem with the expansion.

Yeah, that would be the idea. I would probably create a `:-soup-contains()` and a `:-soup-contains-own()` if we decide to go this direction. We'd probably deprecate `:contains()`. It would kick around for a while to give people a chance to transition, but I think this is the route we'd take if we wanted to expand custom classes further.

That is of course if we don't do something like `:-soup-contains(text <flags>)`. I've tossed around the idea of adding case-insensitivity to `:contains()` for a bit now, and considered doing something similar to what attributes use `[attr=value i]`.

Things to think about.

**Comment by facelessuser** [comment]
All CI issues should now be fixed on the main branch.

@gir-bot add S: needs-decision

**Comment by sm-ghavami** [comment]
> I would probably create a `:-soup-contains()` and a `:-soup-contains-own()` if we decide to go this direction. We'd probably deprecate `:contains()`

Having these kinds of selectors could be useful to target specific elements in the DOM. I don't have any idea about `:contains()` deprecation as it is very self descriptive as a selector.

**Comment by facelessuser** [comment]
I'm not really complaining about how descriptive the selector is. The idea is that things should be consistent and avoid possible conflicts in the future. `:contains-own` is probably descriptive enough as well. `:contains` and `:contains-own` are not part of the CSS spec though, and one day `:contains` or potentially another custom selector could one day be implemented by CSS with a different meaning.

The idea is simply to move non-standard, custom selectors to be namespaced with a browser style prefix.
If I'm potentially opening the door for even more custom selectors, then I run a greater risk of creating the possibility for conflicts with the CSS specification. So, it would be smarter for us to use browser style prefixes much like how Mozilla and Webkit use `:-moz` and `:-webkit` respectively. And if we are doing this for one custom selector, we should do it for all. This way, I don't have to pull the rug out from under anyone if there is a future conflict. I'd rather make the change now with one selector than have to do this later with many.

I think this also serves as a reminder to people who maybe are not as familiar with the CSS specification as well that anything starting with `:-soup` is indeed not standard and not part of the CSS specification.

**Comment by facelessuser** [comment]
@sm-ghavami  I do see one potential issue with this pull. Consider the following:

```html
<div>Test <span>something </span>me</div
```

And if you use the following selector:

```css
:contains-own("test me")
```

This would match as we would extract the content as `Test me`, but this is not quite right as this text is broken up by `something `. We don't want to evaluate `something `, but it is misleading if we are searching for contiguous text. Maybe it should be returned as `["test ", "me"]` and then each part evaluated.

This is simply based on my evaluation of the code, correct me if I am wrong.

**Comment by sm-ghavami** [comment]
> `:contains` and `:contains-own` are not part of the CSS spec though, and one day `:contains` or potentially another custom selector could one day be implemented by CSS with a different meaning.

That's right.
> I think this also serves as a reminder to people who maybe are not as familiar with the CSS specification as well that anything starting with `:-soup` is indeed not standard and not part of the CSS specification.

I got your idea to avoid future conflicts and I think that the `:-soup` prefix could be the right choice for the problem

**Comment by sm-ghavami** [comment]
> This is simply based on my evaluation of the code, correct me if I am wrong.

You're right. I should have noticed this issue before. 
It can be easily fixed by comparing the selector to each part of the content.

File: soupsieve/css_match.py [hunk 1]
```diff
@@ -956,9 +954,20 @@ def match_contains(self, el, contains):
                     content = self.get_text(el, no_iframe=self.is_html)
             found = False
             for text in contain_list.text:
-                if text in content:
-                    found = True
-                    break
+                if type(content) == str:
```
Comment (facelessuser) [inline soupsieve/css_match.py:957]: I'd use `isinstance(content, str)`.

Comment (facelessuser) [reply soupsieve/css_match.py:957]: `isinstance` here as well, or just a simple else as currently we only get a list or a string.

File: soupsieve/css_match.py [hunk 2]
```diff
@@ -956,9 +954,20 @@ def match_contains(self, el, contains):
...
+                    if text in content:
+                        found = True
+                        break
+                elif type(content) == list:
```
Comment (facelessuser) [inline soupsieve/css_match.py:961]: Sorry, I meant for the last comment here. We probably on need an else here as we only have two types of returns we can get here.

File: soupsieve/css_match.py (hunk 1 above)
Comment (sm-ghavami) [reply soupsieve/css_match.py:957]: Nice
I will change the code to follow your code style

File: soupsieve/css_match.py (hunk 2 above)
Comment (sm-ghavami) [reply soupsieve/css_match.py:961]: I know. I tried to keep the function as general as possible :)

**Comment by facelessuser** [comment]
I've created pull #204 which pulls in your work here using the name `:-soup-contains-own()` as well as renaming `:contains()` to `:-soup-contains()`. It is still under consideration, but if it gets merged, it will most likely be merged as implemented in #204.

I removed redundant tests as most of the syntax related tests are already covered by `:contains()`. I specifically targeted tests that I thought maybe illustrated or highlighted `:-soup-contains-own()` specific behavior.

**Comment by facelessuser** [comment]
I've decided to move forward with this as discussed, so in the next release, the functionality will be available as `:-soup-contains-own()`.

**Pull request closed without merging**